### PR TITLE
feat:data shaping

### LINKED ISSUE #197: Data Shaping
Allow the client to select which properties they need

### PR DESCRIPTION
**Pull request by DavidPavlovski**
Implemented data shaping trough query params.
The fields parameter is optional. If left empty we will get all of the properties of the requested resource.

Example: 

`/api/Users/e3171502-622c-4a05-beb8-0c4c37f48401`
The response will be: 

```
{
   "id": "e3171502-622c-4a05-beb8-0c4c37f48401"
   "userName": "exampleUser",
   "email": "email@example.com"
}
```

but if we add the fields parameter: 
`/api/Users/e3171502-622c-4a05-beb8-0c4c37f48401?fields=userName`

The response will only contain the requested properties

```
{
   "userName": "exampleUser"
}
```

Also we can separate field names with comma to select multiple properties: 

`/api/Users/e3171502-622c-4a05-beb8-0c4c37f48401?fields=id,userName`

The response will only contain the requested properties

```
{
   "id": "e3171502-622c-4a05-beb8-0c4c37f48401",
   "userName": "exampleUser"
}
```

closes #197

### CODE REVIEW

File: Dappi.HeadlessCms/Extensions/IEnumerableExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,51 @@
+using System.Dynamic;
+using System.Reflection;
+
+namespace Dappi.HeadlessCms.Extensions
+{
+    public static class IEnumerableExtensions
+    {
+        public static IEnumerable<ExpandoObject> ShapeData<TSource>(this IEnumerable<TSource> source, string? fields)
```
Comment (fvanderflier) [inline Dappi.HeadlessCms/Extensions/IEnumerableExtensions.cs]: Not sure if I follow the logic behind having almost identical lines of code in both extensions? Can't we just re-use the generic or object extension here with using smth like `source.Select(o => o.ShapeObject)`?

File: Dappi.HeadlessCms/Extensions/ObjectExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+            foreach (var field in fields.Split(','))
+            {
+                var propName = field.Trim();
+                var propertyInfo = typeof(TSource).GetProperty(propName,
+                    BindingFlags.IgnoreCase
+                    | BindingFlags.Public
+                    | BindingFlags.Instance);
+                if (propertyInfo is null)
+                {
+                    throw new Exception($"Property {propName} not found in {typeof(TSource).FullName}");
```
[the remark this answers is not in the document]
Comment (DavidPavlovski) [reply Dappi.HeadlessCms/Extensions/ObjectExtensions.cs]: Made a custom exception to handle this. But i think it would be nice to implement an Exception handler middleware and handle exceptions there. Not only will it cleanup our controllers, but it will cleanup the controller template as well, and it will be easier to maintain.

File: Dappi.HeadlessCms/Extensions/EnumerableExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,51 @@
+using System.Dynamic;
+using System.Reflection;
+
+namespace Dappi.HeadlessCms.Extensions
+{
+    public static class IEnumerableExtensions
+    {
+        public static IEnumerable<ExpandoObject> ShapeData<TSource>(this IEnumerable<TSource> source, string? fields)
```
Comment (DavidPavlovski) [reply Dappi.HeadlessCms/Extensions/EnumerableExtensions.cs]: This issue has been resolved. Now instead of using extension methods, we have a DataShaper service, and we are caching the propertyInfos inside a dictionary field. And we only have one implementation of the data shaping logic instead of the two as we had before

**Comment by DavidPavlovski** [comment]
> Like the idea overall. Would suggest some unit tests to find some hidden and nasty edge-cases if they exist.

Unit tests are implemented, we have a lot of coverage for this part. We will see if any bugs arise in the future.

**Review by fvanderflier — APPROVED** [review]
